File: typescript/src/events.ts
import { ProtocolError } from "./errors";

/** A single recognized word with timestamps in seconds. */
export interface Word {
  readonly word: string;
  readonly start: number;
  readonly end: number;
}

/**
 * First event after a successful connection.
 * Wire mapping: `request_id` -> requestId, `model_info` -> modelInfo,
 * `sample_rate` -> sampleRate.
 */
export interface SessionOpened {
  readonly type: "SessionOpened";
  readonly requestId: string;
  readonly modelInfo: { readonly name: string; readonly version: string };
  readonly channels: number;
  readonly sampleRate: number;
  readonly encoding: string;
  /** The event exactly as received from the server. */
  readonly raw: Record<string, unknown>;
}

/**
 * A final transcript delta. Segments are append-only and never revised;
 * concatenating `text` in arrival order reconstructs the transcript verbatim.
 * `start`/`end` are seconds of audio time. All field names match the wire.
 */
export interface Segment {
  readonly type: "Segment";
  readonly channel: number;
  readonly start: number;
  readonly end: number;
  readonly text: string;
  readonly words: readonly Word[];
  /** The event exactly as received from the server. */
  readonly raw: Record<string, unknown>;
}

/** Redaction summary attached to a demand-mode Transcript. */
export interface RedactionSummary {
  readonly applied: boolean;
  readonly policies: readonly string[];
  readonly entitiesRedacted: number;
}

/** A complete demand-mode transcript for one channel. */
export interface Transcript {
  readonly type: "Transcript";
  readonly channel: number;
  readonly text: string;
  readonly words: readonly Word[];
  readonly redaction: RedactionSummary;
  /** The event exactly as received from the server. */
  readonly raw: Record<string, unknown>;
}

/** In-band error codes the server may send. */
export type SttErrorCode =
  | "invalid_params"
  | "invalid_message"
  | "invalid_frame"
  | "idle_timeout"
  | "identity_revalidation_failed"
  | "upstream_unavailable"
  | "transcript_too_large"
  | "internal_error"
  | (string & {});

/** An in-band error event. Field names match the wire. */
export interface ErrorEvent {
  readonly type: "Error";
  readonly code: SttErrorCode;
  readonly message: string;
  /** The event exactly as received from the server. */
  readonly raw: Record<string, unknown>;
}

/**
 * Terminal event of a graceful close. Wire mapping: `request_id` -> requestId,
 * `audio_duration_seconds` -> audioDurationSeconds (the server usage echo),
 * `session_duration_seconds` -> sessionDurationSeconds.
 */
export interface SessionClosed {
  readonly type: "SessionClosed";
  readonly requestId: string;
  readonly audioDurationSeconds: number;
  readonly sessionDurationSeconds: number;
  readonly deliveryFailed: boolean;
  /** The event exactly as received from the server. */
  readonly raw: Record<string, unknown>;
}

/**
 * An event type this SDK version does not recognize. The protocol adds new
 * event types over time; unknown types are surfaced, never treated as errors.
 * `eventType` carries the wire `type` value.
 */
export interface UnknownEvent {
  readonly type: "Unknown";
  readonly eventType: string;
  /** The event exactly as received from the server. */
  readonly raw: Record<string, unknown>;
}

export type SttEvent = SessionOpened | Segment | Transcript | ErrorEvent | SessionClosed | UnknownEvent;

function asObject(value: unknown, context: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new ProtocolError(`${context} is not a JSON object`);
  }
  return value as Record<string, unknown>;
}

function asString(value: unknown, context: string): string {
  if (typeof value !== "string") throw new ProtocolError(`${context} is not a string`);
  return value;
}

function asNumber(value: unknown, context: string): number {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new ProtocolError(`${context} is not a finite number`);
  }
  return value;
}

function asInteger(value: unknown, context: string): number {
  const number = asNumber(value, context);
  if (!Number.isInteger(number)) throw new ProtocolError(`${context} is not an integer`);
  return number;
}

function parseRedaction(raw: Record<string, unknown>): RedactionSummary {
  const redaction = asObject(raw.redaction, "Transcript.redaction");
  if (typeof redaction.applied !== "boolean") {
    throw new ProtocolError("Transcript.redaction.applied is not a boolean");
  }
  if (!Array.isArray(redaction.policies)) {
    throw new ProtocolError("Transcript.redaction.policies is not an array");
  }
  const policies = redaction.policies.map((policy, index) =>
    asString(policy, `Transcript.redaction.policies[${index}]`),
  );
  return {
    applied: redaction.applied,
    policies,
    entitiesRedacted: asInteger(redaction.entities_redacted, "Transcript.redaction.entities_redacted"),
  };
}

/** Parse one server text message into a typed event. Throws ProtocolError. */
export function parseEvent(text: string): SttEvent {
  let value: unknown;
  try {
    value = JSON.parse(text);
  } catch (cause) {
    throw new ProtocolError("server message is not valid JSON", { cause });
  }
  const raw = asObject(value, "server message");
  const type = asString(raw.type, "server message type");
  switch (type) {
    case "SessionOpened": {
      const model = asObject(raw.model_info, "model_info");
      return {
        type,
        requestId: asString(raw.request_id, "request_id"),
        modelInfo: {
          name: asString(model.name, "model_info.name"),
          version: asString(model.version, "model_info.version"),
        },
        channels: asNumber(raw.channels, "channels"),
        sampleRate: asNumber(raw.sample_rate, "sample_rate"),
        encoding: asString(raw.encoding, "encoding"),
        raw,
      };
    }
    case "Segment": {
      const words = raw.words;
      if (!Array.isArray(words)) throw new ProtocolError("Segment.words is not an array");
      return {
        type,
        channel: asNumber(raw.channel, "channel"),
        start: asNumber(raw.start, "start"),
        end: asNumber(raw.end, "end"),
        text: asString(raw.text, "text"),
        words: words.map((entry, index) => {
          const word = asObject(entry, `words[${index}]`);
          return {
            word: asString(word.word, `words[${index}].word`),
            start: asNumber(word.start, `words[${index}].start`),
            end: asNumber(word.end, `words[${index}].end`),
          };
        }),
        raw,
      };
    }
    case "Transcript": {
      const words = raw.words;
      if (!Array.isArray(words)) throw new ProtocolError("Transcript.words is not an array");
      return {
        type,
        channel: asInteger(raw.channel, "channel"),
        text: asString(raw.text, "text"),
        words: words.map((entry, index) => {
          const word = asObject(entry, `words[${index}]`);
          return {
            word: asString(word.word, `words[${index}].word`),
            start: asNumber(word.start, `words[${index}].start`),
            end: asNumber(word.end, `words[${index}].end`),
          };
        }),
        redaction: parseRedaction(raw),
        raw,
      };
    }
    case "Error":
      return {
        type,
        code: asString(raw.code, "code"),
        message: asString(raw.message, "message"),
        raw,
      };
    case "SessionClosed":
      if (raw.delivery_failed !== undefined && typeof raw.delivery_failed !== "boolean") {
        throw new ProtocolError("delivery_failed is not a boolean");
      }
      return {
        type,
        requestId: asString(raw.request_id, "request_id"),
        audioDurationSeconds: asNumber(raw.audio_duration_seconds, "audio_duration_seconds"),
        sessionDurationSeconds: asNumber(raw.session_duration_seconds, "session_duration_seconds"),
        deliveryFailed: raw.delivery_failed ?? false,
        raw,
      };
    default:
      return { type: "Unknown", eventType: type, raw };
  }
}

File: python/src/bw_stt/events.py
"""Typed events and results received from the service."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, TypeAlias

from .errors import ProtocolError

__all__ = [
    "ErrorEvent",
    "Event",
    "RedactedEntity",
    "RedactionSummary",
    "Segment",
    "SessionClosed",
    "SessionOpened",
    "Transcript",
    "Transcription",
    "TranscriptionSegment",
    "UnknownEvent",
    "Word",
    "parse_event",
    "parse_transcription",
]


@dataclass(frozen=True)
class Word:
    """One word piece with timestamps in seconds."""

    word: str
    start: float
    end: float


@dataclass(frozen=True)
class Segment:
    """A final transcript delta; segments are append-only and never revised.

    ``text`` concatenates verbatim across segments: a segment whose text
    starts with a space begins a new word, one without a leading space
    continues the previous word.
    """

    channel: int
    start: float
    end: float
    text: str
    words: tuple[Word, ...]
    raw: dict[str, Any] = field(repr=False)


@dataclass(frozen=True)
class RedactionSummary:
    """PII redaction information attached to a demand-mode transcript."""

    applied: bool
    policies: tuple[str, ...]
    entities_redacted: int


@dataclass(frozen=True)
class RedactedEntity:
    """One PII span returned when redaction entity return is enabled."""

    token: str
    kind: str
    text: str
    start: float | None
    end: float | None


@dataclass(frozen=True)
class Transcript:
    """A complete demand-mode transcript for one channel."""

    channel: int
    text: str
    words: tuple[Word, ...]
    redaction: RedactionSummary
    raw: dict[str, Any] = field(repr=False)
    redacted_entities: tuple[RedactedEntity, ...] | None = None


@dataclass(frozen=True)
class SessionOpened:
    """First event of a session, sent after a successful upgrade."""

    request_id: str
    model_name: str
    model_version: str
    channels: int
    sample_rate: int
    encoding: str
    raw: dict[str, Any] = field(repr=False)


@dataclass(frozen=True)
class SessionClosed:
    """Terminal event of a graceful close; carries the usage echo."""

    request_id: str
    audio_duration_seconds: float
    session_duration_seconds: float
    raw: dict[str, Any] = field(repr=False)
    delivery_failed: bool = False


@dataclass(frozen=True)
class ErrorEvent:
    """An in-band protocol error reported by the server."""

    code: str
    message: str
    raw: dict[str, Any] = field(repr=False)


@dataclass(frozen=True)
class UnknownEvent:
    """An event type this SDK does not know; surfaced for forward compatibility."""

    type: str
    raw: dict[str, Any] = field(repr=False)


@dataclass(frozen=True)
class TranscriptionSegment:
    """One timestamped segment in an offline transcription result."""

    start: float
    end: float
    text: str


@dataclass(frozen=True)
class Transcription:
    """Result of an offline transcribe request."""

    request_id: str
    text: str
    words: tuple[Word, ...]
    segments: tuple[TranscriptionSegment, ...]
    audio_duration_seconds: float
    model_info: dict[str, Any]
    raw: dict[str, Any] = field(repr=False)
    redacted_entities: tuple[RedactedEntity, ...] | None = None


Event: TypeAlias = SessionOpened | Segment | Transcript | ErrorEvent | SessionClosed | UnknownEvent


def _string(obj: dict[str, Any], key: str) -> str:
    value = obj.get(key)
    if not isinstance(value, str):
        raise ProtocolError(f"expected string {key!r} in server message")
    return value


def _integer(obj: dict[str, Any], key: str) -> int:
    value = obj.get(key)
    if isinstance(value, bool) or not isinstance(value, int):
        raise ProtocolError(f"expected integer {key!r} in server message")
    return value


def _number(obj: dict[str, Any], key: str) -> float:
    value = obj.get(key)
    if isinstance(value, bool) or not isinstance(value, int | float):
        raise ProtocolError(f"expected number {key!r} in server message")
    return float(value)


def _optional_number(obj: dict[str, Any], key: str) -> float | None:
    value = obj.get(key)
    if value is None:
        return None
    if isinstance(value, bool) or not isinstance(value, int | float):
        raise ProtocolError(f"expected number or null {key!r} in server message")
    return float(value)


def _boolean(obj: dict[str, Any], key: str, *, default: bool | None = None) -> bool:
    value = obj.get(key, default)
    if not isinstance(value, bool):
        raise ProtocolError(f"expected boolean {key!r} in server message")
    return value


def _words(obj: dict[str, Any]) -> tuple[Word, ...]:
    raw_words = obj.get("words", [])
    if not isinstance(raw_words, list):
        raise ProtocolError("expected array 'words' in server message")
    words = []
    for item in raw_words:
        if not isinstance(item, dict):
            raise ProtocolError("expected object entries in 'words'")
        words.append(Word(_string(item, "word"), _number(item, "start"), _number(item, "end")))
    return tuple(words)


def _redaction(obj: dict[str, Any]) -> RedactionSummary:
    value = obj.get("redaction")
    if not isinstance(value, dict):
        raise ProtocolError("Transcript has no 'redaction' object")
    applied = _boolean(value, "applied")
    raw_policies = value.get("policies")
    if not isinstance(raw_policies, list) or not all(
        isinstance(policy, str) for policy in raw_policies
    ):
        raise ProtocolError("expected string array 'redaction.policies' in server message")
    entities_redacted = value.get("entities_redacted")
    if isinstance(entities_redacted, bool) or not isinstance(entities_redacted, int):
        raise ProtocolError("expected integer 'redaction.entities_redacted' in server message")
    return RedactionSummary(
        applied=applied,
        policies=tuple(raw_policies),
        entities_redacted=entities_redacted,
    )


def _redacted_entities(obj: dict[str, Any]) -> tuple[RedactedEntity, ...] | None:
    if "redacted_entities" not in obj:
        return None
    raw_entities = obj["redacted_entities"]
    if not isinstance(raw_entities, list):
        raise ProtocolError("expected array 'redacted_entities' in server message")
    entities = []
    for item in raw_entities:
        if not isinstance(item, dict):
            raise ProtocolError("expected object entries in 'redacted_entities'")
        entities.append(
            RedactedEntity(
                token=_string(item, "token"),
                kind=_string(item, "kind"),
                text=_string(item, "text"),
                start=_optional_number(item, "start"),
                end=_optional_number(item, "end"),
            )
        )
    return tuple(entities)


def _transcription_segments(obj: dict[str, Any]) -> tuple[TranscriptionSegment, ...]:
    raw_segments = obj.get("segments", [])
    if not isinstance(raw_segments, list):
        raise ProtocolError("expected array 'segments' in transcribe response")
    segments = []
    for item in raw_segments:
        if not isinstance(item, dict):
            raise ProtocolError("expected object entries in 'segments'")
        segments.append(
            TranscriptionSegment(
                start=_number(item, "start"),
                end=_number(item, "end"),
                text=_string(item, "text"),
            )
        )
    return tuple(segments)


def parse_event(payload: str | bytes) -> Event:
    """Decode one server text message into a typed event.

    Unknown event types become :class:`UnknownEvent`; malformed messages
    raise :class:`ProtocolError`.
    """
    try:
        value = json.loads(payload)
    except ValueError as exc:
        raise ProtocolError("server message is not valid JSON") from exc
    if not isinstance(value, dict):
        raise ProtocolError("server message is not a JSON object")
    event_type = value.get("type")
    if not isinstance(event_type, str):
        raise ProtocolError("server message has no 'type'")
    if event_type == "SessionOpened":
        model_info = value.get("model_info")
        if not isinstance(model_info, dict):
            raise ProtocolError("SessionOpened has no 'model_info'")
        return SessionOpened(
            request_id=_string(value, "request_id"),
            model_name=_string(model_info, "name"),
            model_version=_string(model_info, "version"),
            channels=_integer(value, "channels"),
            sample_rate=_integer(value, "sample_rate"),
            encoding=_string(value, "encoding"),
            raw=value,
        )
    if event_type == "Segment":
        return Segment(
            channel=_integer(value, "channel"),
            start=_number(value, "start"),
            end=_number(value, "end"),
            text=_string(value, "text"),
            words=_words(value),
            raw=value,
        )
    if event_type == "Transcript":
        return Transcript(
            channel=_integer(value, "channel"),
            text=_string(value, "text"),
            words=_words(value),
            redaction=_redaction(value),
            raw=value,
            redacted_entities=_redacted_entities(value),
        )
    if event_type == "Error":
        return ErrorEvent(code=_string(value, "code"), message=_string(value, "message"), raw=value)
    if event_type == "SessionClosed":
        return SessionClosed(
            request_id=_string(value, "request_id"),
            audio_duration_seconds=_number(value, "audio_duration_seconds"),
            session_duration_seconds=_number(value, "session_duration_seconds"),
            delivery_failed=_boolean(value, "delivery_failed", default=False),
            raw=value,
        )
    return UnknownEvent(type=event_type, raw=value)


def parse_transcription(payload: str | bytes) -> Transcription:
    """Decode the JSON body of a successful transcribe response."""
    try:
        value = json.loads(payload)
    except ValueError as exc:
        raise ProtocolError("transcribe response is not valid JSON") from exc
    if not isinstance(value, dict):
        raise ProtocolError("transcribe response is not a JSON object")
    model_info = value.get("model_info", {})
    if not isinstance(model_info, dict):
        raise ProtocolError("transcribe response model_info is not an object")
    return Transcription(
        request_id=_string(value, "request_id"),
        text=_string(value, "text"),
        words=_words(value),
        segments=_transcription_segments(value),
        audio_duration_seconds=_number(value, "audio_duration_seconds"),
        model_info=model_info,
        raw=value,
        redacted_entities=_redacted_entities(value),
    )
